### ISSUE TITLE
Feature Request: Implement missing ops from backends

### ISSUE THREAD

**Issue by am17an** [ggml-org/llama.cpp#14909]
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

The list of supported operation in ggml per backend is maintained at https://github.com/ggml-org/llama.cpp/blob/master/docs/ops.md

How to contribute to make the list more complete:

1. Pick a op which is not implemented in a backend (pick ones marked with ❌), pick only one per PR
2. Implement the operation
3. Test your implementation with `test-backend-ops -o <op>`
4. If it passes test for the op, create a PR.

Notes:
1. The ops suffixed with _BACK is training code which is WIP and do not require immediate attention
2. Do not link this issue your PR, which is lead to closing this issue when your PR is merged. Instead, just mention it.

### Motivation

Feature parity between backends is always desirable and it's easy to test the implementation, making it an ideal good first issue.

**Comment by jeemzz147** [comment]
I would like to contribute to this effort. May I ask whether it is okay to work on operators marked as [🟡 Partially supported] in the ops.md table?

**Comment by am17an** [comment]
> I would like to contribute to this effort. May I ask whether it is okay to work on operators marked as [🟡 Partially supported] in the ops.md table?

Yes that would also work. The reason it's preferable to work on completely unsupported ops is because sometimes it's not possible to add full support to an op on a backend (e.g. some quantisation schemes in CUDA) and it might not be clear as a newcomer why that's the case. But if you make an operation support more complete, that would always be welcome!

**Comment by jeemzz147** [comment]
I've verified that PAD_REFLECT_1D is currently unsupported on CUDA.
I would like to work on the PAD_REFLECT_1D op for the CUDA. Please let me know if there are any considerations or edge cases I should keep in mind. Thanks!

**Comment by CISC** [comment]
> I've verified that PAD_REFLECT_1D is currently unsupported on CUDA.

It's already in progress in #14659

**Comment by jeemzz147** [comment]
since `PAD_REFLECT_1D` is already in progress (#14659), I’d like to work on the `SET` op for the CUDA backend instead. Thanks!

**Comment by nimit25** [comment]
Can I pick up CONV_2D for CUDA please? thank you

Do I need to create a separate issue for that?

**Comment by iliailmer** [comment]
Hey!

I'm relatively new to the project, have been using llama.cpp in the past. Really would like to help the project grow and contribute what I can.

Curious about two thing:
1. Is there a good tutorial/guide on getting started with metal framework for this project?
2. I noticed the BLAS is mostly unsupported backend, just. curious to know if there is a fundamental blocking reason behind this?

Thanks in advance!

**Comment by am17an** [comment]
@iliailmer 

1. AFAIK there is no tutorial as such, but `test-backend-ops` always has a reference CPU implementation to compare new implementations with.  Other suggestions would be to look at older PRs added related to the backend you're interested in. 

2. BLAS just forwards mul_mat to another library (see `ggml-blas.cpp`), it would not make sense to add operators there apart from mul_mat and a few others

**Comment by vrvrv** [comment]
Can I pick up CONV_2D for metal?

**Comment by ayush-os** [comment]
Gonna pick up LOG for metal if that's alright?

**Comment by cern1710** [comment]
FYI, I'm gonna pick up OPT_STEP_ADAMW for metal

**Comment by iliailmer** [comment]
i'm working on conv_transpose_2d

**Comment by pra-dan** [comment]
Any reason why nobody is taking up ops for BLAS backend? Literally zero ops =/

**Comment by 0cc4m** [comment]
The BLAS backend is just a way to use a [BLAS](https://en.wikipedia.org/wiki/Basic_Linear_Algebra_Subprograms) library to run matrix multiplications. It's not literally zero ops, `MUL_MAT` is partially supported. Otherwise this "backend" will use the regular CPU operators.

**Comment by codetalker7** [comment]
The referenced table mentions that `CUMSUM` is implemented on the CPU (and some other) backends, but a quick `grep` for `GGML_OP_CUMSUM` leads to no results for me. Is the table incorrect or is `CUMSUM` called something else?

**Comment by 0cc4m** [comment]
A [search](https://github.com/search?q=repo%3Aggml-org%2Fllama.cpp%20GGML_OP_CUMSUM&type=code) does find it.

**Comment by codetalker7** [comment]
My bad; I was searching in my old local directory. Seems that `CUMSUM` is now implemented on CUDA as well. I'll try to work on `NORM_MUL_ADD`.

**Comment by CISC** [comment]
> My bad; I was searching in my old local directory. Seems that `CUMSUM` is now implemented on CUDA as well. I'll try to work on `NORM_MUL_ADD`.

That's a fusion and is only in the list because not all backends have updated their CSVs.

**Comment by CISC** [comment]
@hipudding @noemotiovon @lhez @max-krasnyansky @taronaeo I think CANN, OpenCL and zDNN are the only remaining backends that still have fusion ops in the CSVs, please update them so we can get a more accurate `ops.md`.

**Comment by lhez** [comment]
@CISC Thank you for the reminder. Updated in #17904.

**Comment by taronaeo** [comment]
> @hipudding @noemotiovon @lhez @max-krasnyansky @taronaeo I think CANN, OpenCL and zDNN are the only remaining backends that still have fusion ops in the CSVs, please update them so we can get a more accurate `ops.md`.

@AlekseiNikiforovIBM By any chance can you help with this?

```
test-backend-ops support -b zDNN --output csv
```

**Comment by AlekseiNikiforovIBM** [comment]
Sure, I'll take a look. Although it might take a bit of time at the moment.

**Comment by Old-cpu** [comment]
Hi, Can I contribute to supporting SILU_BACK in Metal?

**Comment by xinpenghuang84-collab** [comment]
I am working on POOL_1D for CPU.

**Comment by engrtipusultan** [comment]
@jeffbolznv @0cc4m I would like to bring to your attention that certain operations are currently unsupported on Vulkan. Implementation of missing ops like #19504  GATED_DELTA_NET etc —in the coming weeks or months, could enhance performance for newer models.

[test-backend-ops.zip](https://github.com/user-attachments/files/25832585/test-backend-ops.zip)

**Comment by 0cc4m** [comment]
I'll look into GATED_DELTA_NET for Vulkan.

**Comment by jeffbolznv** [comment]
@progenyalpha had a prototype of GATED_DELTA_NET at https://github.com/ggml-org/llama.cpp/pull/19504#issuecomment-3911033224, we should give them an opportunity to finish it up.

**Comment by seyoungjeong** [comment]
I'm looking into ADD1 for the Metal backend. Actually nothing calls ADD1 on Metal. Do we need its implementation? @am17an

**Comment by taronaeo** [comment]
> I'm looking into ADD1 for the Metal backend. Actually nothing calls ADD1 on Metal. Do we need its implementation? @am17an 

Looks like someone already submitted a PR: https://github.com/ggml-org/llama.cpp/pull/19987

**Comment by seyoungjeong** [comment]
I'm looking into XIELU for Metal backend.

**Comment by Ssamdeman** [comment]
Hi! I'd love to contribute to this issue as first timer. I'm currently reviewing docs/ops.md to select an unclaimed ❌ op to implement. I'll follow up with my specific op soon.

**Comment by Sero1000** [comment]
Can I pick up Conv_3D for CUDA ?

**Comment by AnandPatil1** [comment]
Hi! I am a student in the CodePath AI301 program and I will be working on implementing the POOL_1D operation for the Vulkan backend.

I have already verified locally using test-backend-ops that it is currently unsupported. I will be working on the implementation and aim to open a PR soon. If anyone has pointers on whether I should closely mirror the existing POOL_2D Vulkan implementation, I would greatly appreciate the guidance!

**Comment by PranavUttarkar** [comment]
Hi I'm also a student in the AI301 program. I'm Picking up Flash Attention EXT and Top K for CUDA. since both those show they are currently unsupported. Opening a PR soon!

**Comment by CISC** [comment]
The docs are not necessarily up-to-date, please update them locally first before proceeding.

Also best to not pick any partially supported ops until you are well acquainted with backends in general (and leave Flash Attention to maintainers). :)

**Comment by soumyadeep-datta** [comment]
Hi! I’m a student in the AI301 program and I’d like to work on implementing GGML_OP_TOP_K for the CUDA backend, which docs/ops.md lists as unsupported (❌) on CUDA. I’ll confirm on master with test-backend-ops -o TOP_K, use the CPU reference for parity, and report perf with test-backend-ops perf -o TOP_K. 

Opening a PR once I’ve verified — happy to coordinate with anyone else looking at TOP_K.

**Comment by Gezahegne** [comment]
Hi! I'm a student in the CodePath AI301 program, and I'd like to work on implementing EXPM1 for the OpenCL backend. I'll start by reproducing the missing backend support locally this week using test-backend-ops -o EXPM1. Any pointers on where to start in the codebase would be appreciated!

**Comment by Gezahegne** [comment]
I see the EXPM1 is already implemented. I'd like to work on implementing ABS for the OpenCL backend. I'll start by reproducing the missing backend support locally this week using test-backend-ops.exe -o ABS.

**Comment by AravindMohan10** [comment]
I’d like to take TOP_K for the ET backend: checked ops.md and searched PRs, doesn’t look like anyone’s on it. Will be my first contribution here, planning to use the CPU impl as reference. Let me know if it’s already spoken for!

**Comment by PranavUttarkar** [comment]
Wanted to request another reviewer for my PR: https://github.com/ggml-org/llama.cpp/pull/25601

**Comment by AravindMohan10** [comment]
> I’d like to take TOP_K for the ET backend: checked ops.md and searched PRs, doesn’t look like anyone’s on it. Will be my first contribution here, planning to use the CPU impl as reference. Let me know if it’s already spoken for!

dug into this and it looks like TOP_K on ET actually needs a proprietary hardware SDK (/opt/et, vendor runtime) that I can't build or test locally, so I'm not able to verify a PR for it. Stepping back from this one!!

**Comment by AravindMohan10** [comment]
> > I’d like to take TOP_K for the ET backend: checked ops.md and searched PRs, doesn’t look like anyone’s on it. Will be my first contribution here, planning to use the CPU impl as reference. Let me know if it’s already spoken for!
> 
> dug into this and it looks like TOP_K on ET actually needs a proprietary hardware SDK (/opt/et, vendor runtime) that I can't build or test locally, so I'm not able to verify a PR for it. Stepping back from this one!!

I'd like to take PAD_REFLECT_1D for the Vulkan backend. Checked ops.md and the PR list : looks open (there was a previous attempt, #25272, but it was withdrawn for AI-disclosure policy reasons, not a technical issue). Let me know if it's already spoken for!

**Comment by Blackcyan30** [comment]
Hi maintainers, I'm new to the repository and contributed support for the SILU_BACK operation on the Metal backend in #25982.
The focused SILU_BACK correctness test and the full Metal backend-op suite pass on my Apple M1 Mac.

**Comment by taronaeo** [comment]
> Hi maintainers, I'm new to the repository and contributed support for the SILU_BACK operation on the Metal backend in #25982.
> The focused SILU_BACK correctness test and the full Metal backend-op suite pass on my Apple M1 Mac.

The Metal maintainers are a little busy as of now. You can occasionally ping @/ggml-org/ggml-metal to do a check-in if the PR was missed.

Do give them some time :)

**Comment by PranavUttarkar** [comment]
Could someone please review: https://github.com/ggml-org/llama.cpp/pull/25601/  @0cc4m @jeffbolznv

**Comment by safiullah3915** [comment]
> I’d like to take TOP_K for the ET backend: checked ops.md and searched PRs, doesn’t look like anyone’s on it. Will be my first contribution here, planning to use the CPU impl as reference. Let me know if it’s already spoken for!

Hey @AravindMohan10 !!! Just wanted to check if you are currently working on PAD_REFLECT_1D for Vulkan or opened a PR for it??? If moved on to another task then I'd love to take it on this !!!

**Comment by Lpcyc** [comment]
Can someone help test these changes ？
My local WSL environment crashed midway through the previous test run, so I wasn't able to complete verification. I'd appreciate it if someone could run the tests and confirm that the kernel works as expected. #26504

**Comment by PranavUttarkar** [comment]
@0cc4m @jeffbolznv I just added Cross entropy loss for Vulkan (https://github.com/ggml-org/llama.cpp/pull/27216) Could you guys please review whenever possible? thanks

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Feature Request: Implement missing ops from backends`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)

**Comment by BensonChen2000** [comment]
Hello I added support for non contiguous unary operation at #28821 I wonder if someone could please review? Thank you.

**Comment by BensonChen2000** [comment]
Hello, I would like to pick up CONV3D for CUDA backend. Thank you.
